SimPEG/potential_fields/base.py

```diff
@@ -73,6 +78,8 @@ def __init__(
         store_sensitivities="ram",
         n_processes=1,
         sensitivity_dtype=np.float32,
+        engine="geoana",
+        numba_parallel=True,
```

Comment:
@santisoler can you explain why we want the `numba_parallel` flag here? I know it determines which version of forward and sensitivity calculation we use (*_parallel or *_serial) but I think it's confusing to control numba parallelism in this way (especially in combination with options `engine = "geoana"` and `n_processes = 1`). I don't think there is ever going to be a performance advantage to `numba_parallel = False`, and if you want to ensure the computation only uses a subset of available threads there are other ways to control that https://numba.pydata.org/numba-doc/latest/user/threading-layer.html

Reply:
Thanks @jedman for the question. The `numba_parallel` flag was moved from the gravity simulation class to the potential fields base simulation class since both the gravity and the magnetic one will use it. This was suggested by @jcapriot in a previous comment and I think it's a good call.

Through that parameter we are deciding if we would like to execute the forward modelling/building the sensitivity matrix functions in parallel or in a single thread. With that simple flag we can control the parallelism _only_ for those methods. The way you point to in Numba's docs applies globally. The goal of the flag is to give more flexibility to control how the simulation should run.

This flexibility might come handy when these simulations are part of a higher order workflow that can be run in parallel, while the simulations will be constrained to a single thread.

The reason to create a separate argument for it different than `n_processes` is to avoid confusion: with `numba_parallel` you can either switch parallelization on or off, but not control in how many threads it should run.

The name of the argument was discussed in #1285, and the reason behind it was to make it clear the difference between the parallelization for the simulation using Numba vs the parallelization of the simulation using multiprocessing.

Would you approach this differently?

Reply:
I totally agree that the base class is the right place to manage parallel execution, but I still think the right way to do this is to use numba's threading control-- if you are running a bunch of BasePFSimulations in a multiprocessing context and you only want numba to use a single thread in each process, then you can set `NUMBA_NUM_THREADS=1` (or use `numba.set_num_threads`). See the example here https://numba.pydata.org/numba-doc/latest/user/threading-layer.html#example-of-limiting-the-number-of-threads

And if the BasePFSimulations are themselves being instantiated in numba threads, you can still control the overall parallelism with numba's threading control. I can't think of a situation where that wouldn't do what you want... 

If it were only up to me, I think the right choice for usability and perofrmance settle on a single "engine" that can be parallelized (or not) with numba, rather than continuing to support multiprocessing, which does not seem to offer a clear advantage. I understand there is a desire to avoid adding to numba to the requirements and maintain backward compatibility, however. 

In any case, if the plan is to keep all these possible code paths, then at a minimum we should update the docstring with an extensive explanation of all the possibilities, and which parameters are related to which other parameters (and when they are ignored). For example, my understanding is that when `engine=choclo` the `n_processes` parameter is ignored

Reply:
> I totally agree that the base class is the right place to manage parallel execution, but I still think the right way to do this is to use numba's threading control-- if you are running a bunch of BasePFSimulations in a multiprocessing context and you only want numba to use a single thread in each process, then you can set NUMBA_NUM_THREADS=1 (or use numba.set_num_threads). See the example here https://numba.pydata.org/numba-doc/latest/user/threading-layer.html#example-of-limiting-the-number-of-threads
> 
> And if the BasePFSimulations are themselves being instantiated in numba threads, you can still control the overall parallelism with numba's threading control. I can't think of a situation where that wouldn't do what you want...

You're right that in that scenario setting the env variable would do just fine. The problem I see with it is that you're changing the Numba behaviour globally. So if another process that uses Numba's parallelization is being used in that workflow, that configuration will change both the PF simulation and this other process.

That's why I mention that whole goal of having `numba_parallel` is to increase the flexibility.

> If it were only up to me, I think the right choice for usability and perofrmance settle on a single "engine" that can be parallelized (or not) with numba, rather than continuing to support multiprocessing, which does not seem to offer a clear advantage. I understand there is a desire to avoid adding to numba to the requirements and maintain backward compatibility, however.

After seeing the improved performance of the Numba-based simulations, we have the future goal of switching to that one permanently in the future. At this stage, since these simulations are among the most used ones by SimPEG users, we considered to take the conservative path of not making breaking changes and allow users to experiment with the new implementation and report bugs they find (as the one that @thibaut-kobold reported in #1373). We aim to make the switch in the future, after we release these implementations.

> In any case, if the plan is to keep all these possible code paths, then at a minimum we should update the docstring with an extensive explanation of all the possibilities, and which parameters are related to which other parameters (and when they are ignored). For example, my understanding is that when engine=choclo the n_processes parameter is ignored

I agree. I probably oversaw the docstrings here, and they should be improved. I'll add more details to them so the different behaviors of the simulation class are well explained. And yes, if `engine="choclo"`, then `n_processes` gets ignored. Users will even get a warning if they are passing both to the simulations' constructors:
https://github.com/simpeg/simpeg/blob/4516cc38439fa2b2defd3539bf991fc3aeefc979/SimPEG/potential_fields/magnetics/simulation.py#L75-L83